src/typechecker.erl

```diff
@@ -1145,8 +1145,10 @@ expect_record_type({type, _, record, [{atom, _, Name}|RefinedTypes]}, Record, #t
                 case RefinedTypes of
                     [] -> Fields;
                     _ ->
-                        [ {typed_record_field, RecordField, RefinedType}
-                        || {{typed_record_field, RecordField, _}, ?type_field_type(_, RefinedType)} <- lists:zip(Fields, RefinedTypes)]
+                        %% `RefinedTypes' and `Fields' might be of different sizes
+                        RefinedMap = maps:from_list([ {N, Ty} || ?type_field_type(N, Ty) <- RefinedTypes ]),
```

Comment:
The name `N` made me think it's a number, but it's a name, right? Please use a different name (such as `FieldName`).